utils/s2n_array.c

```diff
@@ -0,0 +1,105 @@
+/*
+ * Copyright 2016 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License").
+ * You may not use this file except in compliance with the License.
+ * A copy of the License is located at
+ *
+ *  http://aws.amazon.com/apache2.0
+ *
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+#include "utils/s2n_array.h"
+#include "utils/s2n_blob.h"
+#include "utils/s2n_mem.h"
+#include "utils/s2n_safety.h"
+
+#define S2N_INITIAL_ARRAY_SIZE 16
+
+static int s2n_array_embiggen(struct s2n_array *array, uint32_t capacity)
+{
+    struct s2n_blob mem;
+    void *tmp = array->elements;
+
+    GUARD(s2n_alloc(&mem, array->element_size * capacity));
```

Comment:
why can't we realloc() and avoid the zero-then-copy?

you could just zero the newly extended part...

Reply:
Yeah the should work, I was following s2n_map pattern.